src/console_node/nodes.go

```diff
@@ -107,54 +129,151 @@ func pinNumNodes(numAsk, numMax int) int {
 	defer currNodesMutex.Unlock()
 
 	// keep track of if we need to redo the configuration
 	changed := false
 
-	// Update the target number of nodes being monitored
-	updateNodesPerPod()
+	// Figure out how to adjust the number of nodes being monitored
+	deltaMtn, deltaRvr := calcChangeInNodes()
 
-	// Check if we need to gather more nodes - don't take more
-	//  if the service is shutting down
-	if !inShutdown && (len(currentRvrNodes) < targetRvrNodes || len(currentMtnNodes) < targetMtnNodes) {
-		// figure out how many of each to ask for
-		numRvr := pinNumNodes(targetRvrNodes-len(currentRvrNodes), maxAcquireRvr)
-		numMtn := pinNumNodes(targetMtnNodes-len(currentMtnNodes), maxAcquireMtn)
-
-		// attempt to acquire more nodes
-		if numRvr > 0 || numMtn > 0 {
-			// NOTE: this should be the ONLY place where the maps of
-			//  current nodes is updated!!!
-			// NOTE: paradise nodes are included in mountain count
-			newNodes := acquireNewNodes(numMtn, numRvr, podLocData)
-			// process the new nodes
-			for i, node := range newNodes {
-				//log.Printf("  Processing node: %s", node.String())
-				if node.isRiver() {
-					currentRvrNodes[node.NodeName] = &newNodes[i]
-					changed = true
-				} else if node.isMountain() {
-					currentMtnNodes[node.NodeName] = &newNodes[i]
-					changed = true
-				} else if node.isParadise() {
-					currentPdsNodes[node.NodeName] = &newNodes[i]
-					changed = true
-				}
+	// Make sure we can always take one river node if we need to move a worker node
+	if deltaRvr == 0 {
+		deltaRvr = 1
+	}
+
+	log.Printf("doGetNewNodes - deltaRvr: %d, deltaMtn: %d", deltaRvr, deltaMtn)
+
+	// From the change numbers, pull out how many to add (if any)
+	// NOTE: paradise nodes are included in mountain count
+	numAcqRvr := pinNumNodes(deltaRvr, maxAcquireRvr)
+	numAcqMtn := pinNumNodes(deltaMtn, maxAcquireMtn)
+
+	if numAcqRvr > 0 || numAcqMtn > 0 {
+		newNodes := acquireNewNodes(numAcqMtn, numAcqRvr, podLocData)
+		// process the new nodes
+		// NOTE: this should be the ONLY place where the maps of
+		//  current nodes is updated!!!
+		newRvr := 0
+		newMtn := 0
+		newPds := 0
+		for i, node := range newNodes {
+			if node.isRiver() {
+				currentRvrNodes[node.NodeName] = &newNodes[i]
+				changed = true
+				newRvr++
+			} else if node.isMountain() {
+				currentMtnNodes[node.NodeName] = &newNodes[i]
+				changed = true
+				newMtn++
+			} else if node.isParadise() {
+				currentPdsNodes[node.NodeName] = &newNodes[i]
+				changed = true
+				newPds++
 			}
-		} else {
-			log.Printf("Nothing to acquire after pin...")
 		}
-	} else {
-		log.Printf("Skipping acquire - at capacity. CurRvr:%d, TarRvr:%d, CurMtn:%d, TarMtn:%d",
-			len(currentRvrNodes), targetRvrNodes, len(currentMtnNodes)+len(currentPdsNodes), targetMtnNodes)
+		log.Printf(" Added River:%d, Mountain:%d, Paradise:%d", newRvr, newMtn, newPds)
 	}
 
 	// See if we have too many nodes
-	if rebalanceNodes() {
+	if rebalanceNodes(deltaRvr, deltaMtn) {
```

Comment:
So, we have mountain, river, and paradise nodes, but you're only dealing with deltaRvr and deltaMtn. Should there be a deltaParadise or is that a subset of Mountain?

Reply:
Paradise is a subset of mountain. That is documented in a bunch of other places so didn't here.
At some point we will need to make the buckets more generic: 'ipmi' (river), 'ssh w/ key' (mountain), 'ssh w/ creds' (paradise), but that is for another day. Mountain and Paradise are both ssh based so need 'expect' and 'ssh' processes and use the same amount of resources.